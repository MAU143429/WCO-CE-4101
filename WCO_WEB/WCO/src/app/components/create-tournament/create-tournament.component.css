body{
  justify-content: center;
  align-items: center;
  background-color: white;
  height: auto;
  min-height: 100%;

}
.container{
  height: auto;
  max-width: 1400px;
  width: 100%;
  background-color: rgba(255, 255, 255, 0.295);
  padding: 25px 30px;
  border-radius: 5px;
  box-shadow: 0 5px 10px rgba(0,0,0,0.20);
}
.container .title{
  font-size: 30px;
  font-weight: 500;
  text-align: center;
  position: relative;
}
.container .title::before{
  content: "";
  position: absolute;
  bottom: 0;
  height: 3px;
  width: 210px;
  border-radius: 5px;
  background: linear-gradient(135deg, #ffffff, #71aae4, #1277D9);
}
.content form .tournament-details{
  display: flex;
  flex-wrap: wrap;
  margin: 20px 0 12px 0;
}
form .tournament-details .input-box{
  margin-bottom: 15px;
  margin-left: 17%;
  width: calc(50% / 2 );
}
form .tournament-details .input-box1{
  margin-bottom: 15px;
  margin-left: 38%;
  width: calc(23%);
}
form .tournament-details .input-box2{
  margin-bottom: 15px;
  margin-left: 32%;
  width: calc(35% - 20px);
}
form .input-box span.details{
  display: block;
  font-weight: 500;
  margin-bottom: 5px;
  text-align: center;
}

form .input-box1 span.details{
  display: block;
  font-weight: 500;
  margin-bottom: 5px;
  text-align: center;
}

form .input-box2 span.details{
  display: block;
  font-weight: 500;
  margin-bottom: 5px;
  text-align: center;
}
.my-check{
  margin-left: 40%;

}
.tournament-details .input-box input{
  height: 45px;
  width: 100%;
  font-size: 16px;
  border-radius: 5px;
  padding-left: 15px;
  border: 1px solid #ccc;
  border-bottom-width: 2px;
  transition: all 0.3s ease;
}
.tournament-details .input-box1 input{
  height: 45px;
  width: 100%;
  font-size: 16px;
  border-radius: 5px;
  padding-left: 15px;
  border: 1px solid #ccc;
  border-bottom-width: 2px;
  transition: all 0.3s ease;
}

.textarea{
  height: 200px;
  width: 100%;
  font-size: 16px;
  border-radius: 5px;
  padding-left: 15px;
  border: 1px solid #ccc;
  border-bottom-width: 2px;
  transition: all 0.3s ease;
}

 form .category{
   display: flex;
   width: 100%;
   margin-left: 10%;
   margin: 14px 0 ;

 }
 form .category label{
   display: flex;
   align-items: center;
   cursor: pointer;
   margin-right: 15px;
 }
 form .category label .dot{
  height: 18px;
  width: 18px;
  border-radius: 50%;
  margin-right: 15px;
  background: #d9d9d9;
  border: 5px solid transparent;
  transition: all 0.3s ease;
}
 #dot-1:checked ~ .category label .one,
 #dot-2:checked ~ .category label .two{
   background: #1277D9;
   border-color: #d9d9d9;

 }
 form input[type="radio"]{
   display: none;

 }
 form .button{
   height: 45px;
   margin: 35px 0
 }

 /* Aspecto del boton de crear torneo */
 form .button input{
  height: 100%;
  width: 40%;
  margin-left: 30%;
  border-radius: 5px;
  border: none;
  color: #fff;
  font-size: 18px;
  font-weight: 500;
  letter-spacing: 1px;
  cursor: pointer;
  transition: all 0.3s ease;
  background: linear-gradient(135deg, #1277D9, #7bb0e6, #1277D9);
}
 form .button input:hover{
  background: linear-gradient(135deg, #1277D9, #7bb0e6, #1277D9);
  }
 @media(max-width: 584px){
 .container{
  max-width: 100%;
}

}

.mat-button{
  background-color:#1277D9;
  color: white;
}

.mat-table{
  width: 100%;
  background-color:white ;
}

.mat-header-cell,
.mat-cell {
  color: #1277D9;
}

.popup{
  width: 500px;
  height: 252px;
  background-size:cover;
  background-position: center;
}










