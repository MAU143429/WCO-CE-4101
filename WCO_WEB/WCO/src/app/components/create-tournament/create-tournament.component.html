<!DOCTYPE html>
<html lang="en" dir="ltr">
  <head>
    <meta charset="UTF-8">
    <!---Referencia al archivo css que contiene el estilo--->
    <link rel="stylesheet" href="create-tournament.component.css">
     <meta name="viewport" content="width=device-width, initial-scale=1.0">
   </head>
<!---Componentes que conforman la seccion de crear un torneo--->

<body>
  <app-navbar></app-navbar>
  <br>
  <br>
  <br>
  <br>
  <div class="container">
    <div class="title">Crear un torneo</div>
    <div class="content">
      <!---Formulario donde se reciben la informacion para crear el torneo--->
      <form>
        <div class="tournament-details">
          <div class="input-box1">
            <span class="details" >Nombre del torneo<span style="color: rgb(245, 59, 59);">*</span> </span>
            <input type="text" [(ngModel)]= "newTournament.name" minlength="5" maxlength="30" [ngModelOptions]="{standalone: true}"  placeholder="Ingrese un nombre para el torneo aquí..." required>
          </div>
          <div class="input-box">
            <span class="details">Fecha de inicio<span style="color: rgb(245, 59, 59);">*</span> </span>
            <input type="date"  [(ngModel)]= "newTournament.startDate" [ngModelOptions]="{standalone: true}" required>
          </div>
          <div class="input-box">
            <span class="details"  >Fecha de final<span style="color: rgb(245, 59, 59);">*</span> </span>
            <input type="date"  [(ngModel)]= "newTournament.endDate" [ngModelOptions]="{standalone: true}"  required>
          </div>
          <div class="input-box2">
            <span class="details" >Descripción del torneo </span>
            <textarea class="textarea" minlength="5" maxlength="1000" [(ngModel)]= "newTournament.description" [ngModelOptions]="{standalone: true}"  placeholder="Ingrese una descripción del torneo aquí.." ></textarea>
          </div>
          <div class="input-box1">
            <span class="details"  >Seleccione el tipo de torneo <span style="color: rgb(245, 59, 59);">*</span></span>
            <div class="details" >
              <input type="radio"  id="dot-1" [(ngModel)]="radioSelected" name="list_name" value="Selecciones" (change)="onItemChange()">
              <input type="radio"  id="dot-2" [(ngModel)]="radioSelected" name="list_name" value="Equipos Locales" (change)="onItemChange()">
              <div class="category">
                <label for="dot-1">
                <span class="dot one"></span>
                <span class="rol">Selecciones</span>
              </label>
              <label for="dot-2">
                <span class="dot two"></span>
                <span class="rol">Equipos Locales</span>
              </label>
              </div>
            </div>
          </div>
          <div class="input-box">
            <span class="details">Equipos del torneo<span style="color: rgb(245, 59, 59);">*</span></span>
            <mbsc-page>
              <mbsc-select [data]="dropdownData" [(ngModel)]="newTournament.teams" [ngModelOptions]="{standalone: true}" [selectMultiple]="true"  labelStyle="stacked"  placeholder="Seleccione los equipos..."></mbsc-select>
            </mbsc-page>
          </div>

          <div class="input-box">
            <span class="details">Fases del torneo<span style="color: rgb(245, 59, 59);">*  </span><button mat-button (click)="openDialog('Agregar',{})" >Agregar Fase</button></span>


            <table mat-table [dataSource]="bracketSource" #mytable class="my-table mat-elevation-z8">

              <ng-container matColumnDef="name">
                <th mat-header-cell *matHeaderCellDef> Nombre de la fase </th>
                <td mat-cell *matCellDef="let element"> {{element.name}} </td>
              </ng-container>

              <ng-container matColumnDef="action">
                <th mat-header-cell *matHeaderCellDef> Opciones </th>
                <td mat-cell *matCellDef="let element" class="action-link">
                  <a (click)="openDialog('Eliminar',element)"> Eliminar</a>
                </td>
              </ng-container>

              <tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
              <tr mat-row *matRowDef="let row; columns: displayedColumns;"></tr>
            </table>


          </div>
        </div>
        <div class="button">
          <input type="submit" value="Crear Torneo" (click)="addTournament()">
        </div>
      </form>
    </div>
  </div>






</body>

<ng-template #content let-modal>
  <div class="popup">
    <button class="close cbtn" aria-label="Close" (click)="modal.dismiss('Cross click')"></button>
  </div>
</ng-template>

</html>
