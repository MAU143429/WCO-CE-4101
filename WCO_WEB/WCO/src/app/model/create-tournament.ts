import { DbBracket } from './db-bracket';
import { DbTeam } from './db-team';

export class CreateTournament {
  name: string = '';
  startDate: string = '';
  endDate: string = '';
  type: string = '';
  description: string = '';
  teams: DbTeam[] = [];
  brackets: String[] = [];
}
