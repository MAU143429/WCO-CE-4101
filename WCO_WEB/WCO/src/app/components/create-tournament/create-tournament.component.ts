import { Component, OnInit, ViewChild } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { MatTable } from '@angular/material/table';
import { DialogBoxComponent } from '../dialog-box/dialog-box.component';
import { CreateTournament } from '../../model/create-tournament';
import { setOptions, localeEs } from '@mobiscroll/angular';
import { TournamentService } from '../../services/tournament.service';
import { Teams } from '../../interface/teams';

export interface bracketName {
  name: string;
}

export interface Type {
  type: string;
}

export const ITEMS: Type[] = [
  { type: 'Selecciones' },
  { type: 'Equipos Locales' },
];

setOptions({
  locale: localeEs,
  themeVariant: 'light',
});

@Component({
  selector: 'app-create-tournament',
  templateUrl: './create-tournament.component.html',
  styleUrls: ['./create-tournament.component.css'],
})
export class CreateTournamentComponent implements OnInit {
  /**Instanciando variables del sistema */

  newTournament: CreateTournament = new CreateTournament();
  radioSel: any;
  radioSelected: string = '';
  Option: Type[] = ITEMS;
  dropdownData: Teams[] = [];
  displayedColumns: string[] = ['name', 'action'];
  bracketSource: bracketName[] = [];
  selectedTeams: string[] = [];

  @ViewChild(MatTable, { static: true }) table: MatTable<any> | any;
  constructor(public dialog: MatDialog, private service: TournamentService) {}

  /**
   * @brief Este metodo permite obtener el valor seleccionado en el radio button y realizar alguna acciones
   * relacionadas a la seleccion dada por el usuario
   */
  getSelecteditem() {
    this.radioSel = ITEMS.find((Item) => Item.type === this.radioSelected);
    this.newTournament.type = this.radioSelected;
    if (this.radioSelected == 'Selecciones') {
      this.service.getTeams().subscribe((data) => (this.dropdownData = data));
    } else {
      this.service.getCFTeams().subscribe((data) => (this.dropdownData = data));
    }
  }

  /**
   * @brief Este metodo detecta cuando existe un cambio en el radio button
   */
  onItemChange() {
    this.getSelecteditem();
  }

  setBracketList() {
    this.bracketSource.forEach((value) => {
      this.newTournament.brackets.push(value.name);
    });
  }

  addTournament() {
    this.setBracketList();

    this.service
      .setTournament(this.newTournament)
      .subscribe((tournament) => console.log(this.newTournament));
  }

  ngOnInit(): void {}

  openDialog(action: String, obj: any) {
    obj.action = action;
    const dialogRef = this.dialog.open(DialogBoxComponent, {
      width: '250px',
      data: obj,
      panelClass: 'custom-modalbox',
    });

    dialogRef.afterClosed().subscribe((result) => {
      if (result.event == 'Agregar') {
        this.addRowData(result.data);
      } else if (result.event == 'Eliminar') {
        this.deleteRowData(result.data);
      }
    });
  }

  addRowData(row_obj: any) {
    this.bracketSource.push({
      name: row_obj.name,
    });
    this.table.renderRows();
  }

  deleteRowData(row_obj: any) {
    this.bracketSource = this.bracketSource.filter((value, key) => {
      return value.name != row_obj.name;
    });
  }
}
