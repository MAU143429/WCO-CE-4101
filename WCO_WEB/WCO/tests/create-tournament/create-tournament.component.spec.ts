import { ComponentFixture, TestBed } from '@angular/core/testing';
import { CreateTournamentComponent } from '../../src/app/components/create-tournament/create-tournament.component';

import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing'
import { MatSnackBar, MatSnackBarModule } from '@angular/material/snack-bar';
import { ModalDismissReasons } from '@ng-bootstrap/ng-bootstrap';
import { TournamentService } from '../../src/app/services/tournament.service';
import { Router } from '@angular/router';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { BrowserAnimationsModule, NoopAnimationsModule } from '@angular/platform-browser/animations';
import { NoopAnimationPlayer } from '@angular/animations';
import { RouterTestingModule } from '@angular/router/testing';
import { Observable, of } from 'rxjs';
import { MatDialogModule } from '@angular/material/dialog';
import fi from '@mobiscroll/angular/dist/js/i18n/fi';

const dummyTournament = [{
  name: 'Mundial',
  startDate: '10/10/2022',
  endDate: '11/11/2022',
  type: 'Selecciones',
  description: 'Woww mundial',
  teams: [{name: "T1", id: 1}, {name: "TB", id: 2}],
  brackets: ["Final"]
}]
const mockTournamentService: {
  getTeamsbyType: () => Observable<any>,
  setTournament: () => Observable<any>
} = {
  getTeamsbyType: () => of([{name: "T1", id: 1}, {name: "TB", id: 2}]),
  setTournament: () => of(dummyTournament)
};


describe('CreateTournamentComponent', () => {
  let component: CreateTournamentComponent;
  let fixture: ComponentFixture<CreateTournamentComponent>;
  let compiled: HTMLElement;
  let httpMock: HttpTestingController;
  let service: TournamentService;
  let router: Router;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ CreateTournamentComponent ],
      imports: [ 
        HttpClientTestingModule, 
        MatSnackBarModule,
        //FormsModule,
        //ReactiveFormsModule,
        MatDialogModule,
        NoopAnimationsModule
      ],
      providers: [
        {provide: TournamentService, useValue: mockTournamentService}
      ]
    })
    .compileComponents();
    
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(CreateTournamentComponent);
    component = fixture.componentInstance;
    service = TestBed.inject( TournamentService );
    httpMock = TestBed.inject( HttpTestingController );
    compiled = fixture.nativeElement;
    fixture.detectChanges();

  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe de hace match con el snapshot', () => {
    expect(compiled.innerHTML).toMatchSnapshot();
  });

  test("getSelectedItem()", () => {
    component.radioSelected = 'Selecciones';

    const getTeamsSpy = jest.spyOn(mockTournamentService, "getTeamsbyType")
    getTeamsSpy.mockReturnValue(of([{name: "T1", teId: 1}, {name: "TB", teId: 2}]));

    component.getSelecteditem();
    expect(component.dropdownData).toEqual([{text: "T1", value: 1}, {text: "TB", value: 2}])
  })

  test('deberia obtener nombres de brackets en formato correcto para API usando setBracketList()', () =>{
    component.bracketSource = [{name: "Final"}]
    component.setBracketList();
    fixture.detectChanges()

    expect(component.newTournament.brackets).toEqual(["Final"]);
  })

  test('Se deberia agregar torneo, addTournament()', () => {

    jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));

    const setTournamentSpy = jest.spyOn(mockTournamentService, 'setTournament')
    setTournamentSpy.mockReturnValue(of(dummyTournament));

    const errorSpy = jest.spyOn( component, 'openError')

    component.newTournament.endDate = '11/11/2022';
    component.newTournament.name = 'Mundial';
    component.newTournament.startDate = '10/10/2022';
    component.newTournament.teams.length = 2;
    component.newTournament.brackets.length = 1;

    component.addTournament();
    expect(setTournamentSpy).toHaveBeenCalled();
  })

  test('Error por fecha inicio depues de fecha fin, addTournament()', () => {

    jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));

    const setTournamentSpy = jest.spyOn(mockTournamentService, 'setTournament')
    setTournamentSpy.mockReturnValue(of(dummyTournament));

    const errorSpy = jest.spyOn( component, 'openError')

    component.newTournament.endDate = '11/11/2022';
    component.newTournament.name = 'Mundial';
    component.newTournament.startDate = '12/12/2022';
    component.newTournament.teams.length = 2;
    component.newTournament.brackets.length = 1;

    component.addTournament();
    expect(errorSpy).toHaveBeenCalledWith('Error en fechas ingresadas!', 'Fecha final incorrecta!');
  })

  test('Error por falta de equipos, addTournament()', () => {

    jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));

    const setTournamentSpy = jest.spyOn(mockTournamentService, 'setTournament')
    setTournamentSpy.mockReturnValue(of(dummyTournament));

    const errorSpy = jest.spyOn( component, 'openError')

    component.newTournament.endDate = '11/11/2022';
    component.newTournament.name = 'Mundial';
    component.newTournament.startDate = '10/10/2022';
    component.newTournament.teams.length = 1;
    component.newTournament.brackets.length = 1;

    component.addTournament();
    expect(errorSpy).toHaveBeenCalledWith('Se requiere al menos dos equipos para crear un torneo!',
    'Ingrese algÃºn otro antes de continuar!');
  })

  test('Error por fecha inicio en el pasado, addTournament()', () => {

    jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));

    const setTournamentSpy = jest.spyOn(mockTournamentService, 'setTournament')
    setTournamentSpy.mockReturnValue(of(dummyTournament));

    const errorSpy = jest.spyOn( component, 'openError')

    component.newTournament.endDate = '11/11/2022';
    component.newTournament.name = 'Mundial';
    component.newTournament.startDate = '12/12/2019';
    component.newTournament.teams.length = 2;
    component.newTournament.brackets.length = 1;

    component.addTournament();
    expect(errorSpy).toHaveBeenCalledWith('Error no se pueden generar torneos en el pasado!',
    'Ingrese una fecha correcta para continuar!');
  })

  test('Inputs funcionan', () => {

    const nameInput1: HTMLInputElement = fixture.debugElement.nativeElement.querySelector('#nameInput');
    nameInput1.value = 'Mundial 1';
    nameInput1.dispatchEvent(new Event('input'));

    expect(component.newTournament.name == nameInput1.value)

    const startDateInput: HTMLInputElement = fixture.debugElement.nativeElement.querySelector('#startDateInput');
    startDateInput.value = '10/10/2022'
    startDateInput.dispatchEvent(new Event('input'));

    expect(component.newTournament.startDate).toEqual(startDateInput.value)

    const endDateInput: HTMLInputElement = fixture.debugElement.nativeElement.querySelector('#endDateInput');
    endDateInput.value = '11/11/2022'
    endDateInput.dispatchEvent(new Event('input'));

    expect(component.newTournament.endDate).toEqual(endDateInput.value)

    const dot1: HTMLInputElement = fixture.debugElement.nativeElement.querySelector('#dot-1');
    dot1.dispatchEvent(new Event( 'click' ))

    expect(component.newTournament.type == dot1.value)
  })

});
