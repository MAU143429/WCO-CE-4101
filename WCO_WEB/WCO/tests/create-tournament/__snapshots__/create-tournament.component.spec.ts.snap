// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTournamentComponent debe de hace match con el snapshot 1`] = `"<html lang=\\"en\\" dir=\\"ltr\\" ng-reflect-dir=\\"ltr\\"><head><meta charset=\\"UTF-8\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\"></head><body><app-navbar></app-navbar><br><br><br><br><div class=\\"container\\"><div class=\\"title\\">Crear un torneo</div><div class=\\"content\\"><form><div class=\\"tournament-details\\"><div class=\\"input-box1\\"><span class=\\"details\\">Nombre del torneo<span style=\\"color: rgb(245, 59, 59);\\">*</span></span><input type=\\"text\\" minlength=\\"5\\" maxlength=\\"30\\" placeholder=\\"Ingrese un nombre para el torneo aquí...\\" id=\\"nameInput\\" required=\\"\\"></div><div class=\\"input-box\\"><span class=\\"details\\">Fecha de inicio<span style=\\"color: rgb(245, 59, 59);\\">*</span></span><input type=\\"date\\" id=\\"startDateInput\\" required=\\"\\"></div><div class=\\"input-box\\"><span class=\\"details\\">Fecha de final<span style=\\"color: rgb(245, 59, 59);\\">*</span></span><input type=\\"date\\" id=\\"endDateInput\\" required=\\"\\"></div><div class=\\"input-box2\\"><span class=\\"details\\">Descripción del torneo </span><textarea minlength=\\"5\\" maxlength=\\"1000\\" placeholder=\\"Ingrese una descripción del torneo aquí..\\" class=\\"textarea\\"></textarea></div><div class=\\"input-box1\\"><span class=\\"details\\">Seleccione el tipo de torneo <span style=\\"color: rgb(245, 59, 59);\\">*</span></span><div class=\\"details\\"><input type=\\"radio\\" id=\\"dot-1\\" name=\\"list_name\\" value=\\"Selecciones\\"><input type=\\"radio\\" id=\\"dot-2\\" name=\\"list_name\\" value=\\"Equipos Locales\\"><div class=\\"category\\"><label for=\\"dot-1\\"><span class=\\"dot one\\"></span><span class=\\"rol\\">Selecciones</span></label><label for=\\"dot-2\\"><span class=\\"dot two\\"></span><span class=\\"rol\\">Equipos Locales</span></label></div></div></div><div class=\\"input-box\\"><span class=\\"details\\">Equipos del torneo<span style=\\"color: rgb(245, 59, 59);\\">*</span></span><mbsc-page><mbsc-select labelstyle=\\"stacked\\" placeholder=\\"Seleccione los equipos...\\"></mbsc-select></mbsc-page></div><div class=\\"input-box\\"><span class=\\"details\\">Fases del torneo<span style=\\"color: rgb(245, 59, 59);\\">* </span><button mat-button=\\"\\">Agregar Fase</button></span><table mat-table=\\"\\" class=\\"my-table mat-elevation-z8\\"><!--container--><!--container--><!--ng-container--><!--container--><!--container--><!--ng-container--><!--container--><!--container--></table></div></div><div class=\\"button\\"><input type=\\"submit\\" value=\\"Crear Torneo\\" routerlink=\\"/home\\"></div></form></div></div></body></html>"`;
